tests/v2/validation/provisioning/k3s/encryption_key_rotation_test.go

```diff
@@ -0,0 +1,246 @@
+
+	kubeProvisioningClient, err := client.GetKubeAPIProvisioningClient()
+	require.NoError(t, err)
+
+	cluster, err := client.Steve.SteveType(clusters.ProvisioningSteveResourceType).ByID(steveID)
+	require.NoError(t, err)
+
+	clusterSpec := &apiv1.ClusterSpec{}
+	err = v1.ConvertToK8sType(cluster.Spec, clusterSpec)
+	require.NoError(t, err)
+
+	updatedCluster := *cluster
+
+	clusterSpec.RKEConfig.RotateEncryptionKeys = &rkev1.RotateEncryptionKeys{
+		Generation: generation,
+	}
+
+	updatedCluster.Spec = *clusterSpec
+
+	cluster, err = client.Steve.SteveType(clusters.ProvisioningSteveResourceType).Update(cluster, updatedCluster)
+	require.NoError(t, err)
+
+	for _, phase := range phases {
+		err = kwait.Poll(10*time.Second, timeout, IsAtLeast(t, client, namespace, cluster.ObjectMeta.Name, phase))
+		require.NoError(t, err)
+	}
+
+	clusterWait, err := kubeProvisioningClient.Clusters(namespace).Watch(context.TODO(), metav1.ListOptions{
+		FieldSelector:  "metadata.name=" + cluster.ObjectMeta.Name,
+		TimeoutSeconds: &defaults.WatchTimeoutSeconds,
+	})
+	require.NoError(t, err)
+
+	err = wait.WatchWait(clusterWait, clusters.IsProvisioningClusterReady)
+	require.NoError(t, err)
+
+	t.Logf("Successfully completed encryption key rotation for %s", cluster.ObjectMeta.Name)
+}
+
+func createSecretsForCluster(t *testing.T, client *rancher.Client, steveID string, scale int) {
+	t.Logf("Creating %d secrets in namespace default for encryption key rotation", scale)
+
+	_, clusterName, found := strings.Cut(steveID, "/")
+	require.True(t, found)
+
+	clusterID, err := clusters.GetClusterIDByName(client, clusterName)
+	require.NoError(t, err)
+	secretResource, err := kubeapi.ResourceForClient(client, clusterID, "default", secrets.SecretGroupVersionResource)
+	require.NoError(t, err)
+
+	for i := 0; i < scale; i++ {
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: fmt.Sprintf("encryption-key-rotation-test-%d-", i),
+			},
+			Data: map[string][]byte{
+				"key": []byte(namegen.RandStringLower(5)),
+			},
+		}
+		_, err = secrets.CreateSecret(secretResource, secret)
+		require.NoError(t, err)
+	}
+}
+
+func (r *K3SEncryptionKeyRotationTestSuite) TestEncryptionKeyRotation() {
+	for _, providerName := range r.config.Providers {
+		for _, kubernetesVersion := range r.config.K3SKubernetesVersions {
+			// cleanup resources inside the for loop to prevent leaking
+			subSession := r.session.NewSession()
+
+			provider := CreateProvider(providerName)
+
+			client, err := r.client.WithSession(subSession)
+			require.NoError(r.T(), err)
+
+			cloudCredential, err := provider.CloudCredFunc(client)
+			require.NoError(r.T(), err)
+
+			// provisioning is not considered part of the test
+			id := provisionEnvironment(r.T(), client, r.clusterName, provider, kubernetesVersion, r.config.NodesAndRoles, cloudCredential, r.config.PSACT, r.config.AdvancedOptions)
```

Comment:
If provisioning is not considered a part of the test would it make sense to have it in its own test package and receive the cluster ID like we do in other tests?
ex) https://github.com/rancher/rancher/blob/a05de31fccb10059447c169f28dcc2068982a6f0/tests/v2/validation/charts/gatekeeper_test.go#L42-L47

Reply:
This way you don't have to duplicate test for k3s as well.